Mathlib/RepresentationTheory/Coinvariants.lean

```diff
@@ -0,0 +1,182 @@
+
+@[simp]
+theorem lift_comp_mk (f : V →ₗ[k] W) (h : ∀ (x : G), f ∘ₗ ρ x = f) :
+    lift ρ f h ∘ₗ mk ρ = f := rfl
+
+@[simp]
+theorem lift_mk (f : V →ₗ[k] W) (h : ∀ (x : G), f ∘ₗ ρ x = f) (x : V) :
+    lift ρ f h (mk _ x) = f x := rfl
+
+@[ext]
+lemma hom_ext {f g : Coinvariants ρ →ₗ[k] W} (H : f ∘ₗ mk ρ = g ∘ₗ mk ρ) : f = g :=
+  Submodule.linearMap_qext _ H
+
+variable (ρ)
+
+/-- Given `G`-representations on `k`-modules `V, W`, a linear map `V →ₗ[k] W` commuting with
+the representations induces a `k`-linear map between the coinvariants. -/
+noncomputable def map (f : V →ₗ[k] W) (hf : ∀ g, f ∘ₗ ρ g = τ g ∘ₗ f) :
+    Coinvariants ρ →ₗ[k] Coinvariants τ :=
+  lift _ (mk _ ∘ₗ f) fun g => LinearMap.ext fun x => (mk_eq_iff _).2 <|
+    mem_augmentationSubmodule_of_eq (ρ := τ) g (f x) _ <| by simpa using congr($((hf g).symm) x)
+
+variable {ρ τ}
+
+@[simp]
+lemma map_comp_mk {f : V →ₗ[k] W} {hf : ∀ g, f ∘ₗ ρ g = τ g ∘ₗ f} :
+    map ρ τ f hf ∘ₗ mk ρ = mk τ ∘ₗ f := rfl
+
+@[simp]
+lemma map_mk (f : V →ₗ[k] W) (hf : ∀ g, f ∘ₗ ρ g = τ g ∘ₗ f) (x : V) :
+    map ρ τ f hf (mk _ x) = mk _ (f x) := rfl
+
+end Coinvariants
+end Representation
+
+namespace Rep
+
+open CategoryTheory Representation
+
+namespace Coinvariants
+
+variable {k G : Type u} [CommRing k] [Monoid G] {A B C : Rep k G} {n : ℕ}
+
+/-- The linear map underlying a `G`-representation morphism `A ⟶ B`, where `B` has the trivial
+representation, factors through `A_G`. -/
+noncomputable abbrev lift [B.ρ.IsTrivial] (f : A ⟶ B) :
+    A.ρ.Coinvariants →ₗ[k] B :=
+  Representation.Coinvariants.lift _ f.hom.hom fun _ => by
+    ext
+    have := hom_comm_apply f
+    simp_all
+
+/-- A `G`-representation morphism `A ⟶ B` induces a linear map `A_G →ₗ[k] B_G`. -/
+noncomputable abbrev map (f : A ⟶ B) :
+    A.ρ.Coinvariants →ₗ[k] B.ρ.Coinvariants :=
+  Representation.Coinvariants.map _ _ f.hom.hom fun _ => ModuleCat.hom_ext_iff.1 (f.comm _)
+
+@[simp]
+theorem map_id (A : Rep k G) :
+    map (𝟙 A) = LinearMap.id := by
+  ext; rfl
+
+@[simp]
+theorem map_comp (f : A ⟶ B) (g : B ⟶ C) :
+    map (f ≫ g) = map g ∘ₗ map f := by
+  ext; rfl
+
+variable (k G)
+
+/-- The functor sending a representation to its coinvariants. -/
+@[simps obj map]
+noncomputable def functor : Rep k G ⥤ ModuleCat k where
+  obj A := ModuleCat.of k (A.ρ.Coinvariants)
+  map f := ModuleCat.ofHom (map f)
+
+variable {k G} in
+@[ext]
+lemma functor_hom_ext {M : ModuleCat k} {f g : (functor k G).obj A ⟶ M}
+    (hfg : f.hom ∘ₗ Coinvariants.mk A.ρ = g.hom ∘ₗ Coinvariants.mk A.ρ) :
+    f = g := ModuleCat.hom_ext <| Coinvariants.hom_ext hfg
+
+instance : (functor k G).Additive where
+  map_add := ModuleCat.hom_ext <| LinearMap.ext fun x => Quotient.inductionOn' x (fun _ => rfl)
+
+/-- The adjunction between the functor sending a representation to its coinvariants and the functor
+equipping a module with the trivial representation. -/
+noncomputable def adjunction : functor k G ⊣ trivialFunctor G :=
+  Adjunction.mkOfHomEquiv {
```

Comment:
You might want to use the usual constructor if you care about the defers on units and counits.